Update post author block functionality and visual parity

Addison-Stavlo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
Update the post author block functionality as issued in #21087.

This introduces:
* <s>Author font size</s>
* <s>Byline font size</s>
* <s>Bio font size</s>
* __experimentalFontSize support
* Relative sizing for byline and bio fields.
* Remaining Rich Text capabilities for the byline.

This also fixes issues with:
* Text alignment
* Font and Background Colors
* Default attributes
* Overall parity between editors and front-end

![editing-post-author](https://user-images.githubusercontent.com/28742426/83816432-5df41480-a690-11ea-98ea-efd1db461195.gif)

### A note on front-end parity:
Preset font sizes ( 'Normal', 'Large', etc. ) run on class names that may have styles added by themes.  For example, TwentyTwenty adds styles such as:
```
.entry-content .has-large-font-size {
    font-size: 1.25em;
    line-height: 1.4;
}
```
This overwrites the standard definition for large text on the front end and causes it to be a different size than the expected 36px as is shown in the editor.  Im not sure what we can/should do about this, as if we force parity on this we disable the themes ability to have control over preset sizing.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
Tested on local docker environment.
Test various settings, save, and verify visual parity between site editor, post editor, and front end.

## Screenshots <!-- if applicable -->
![Screen Shot 2020-06-04 at 7 05 32 PM](https://user-images.githubusercontent.com/28742426/83818896-72d3a680-a696-11ea-895d-5ab57eac989b.png)

![Screen Shot 2020-06-04 at 8 16 14 PM](https://user-images.githubusercontent.com/28742426/83822954-8d128200-a6a0-11ea-9c05-85024719b482.png)

## Types of changes
<!-- What types of changes does your code introduce?  -->
Bug fix (non-breaking change which fixes an issue)
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Addison-Stavlo committed:
added more font sizes to control

Addison-Stavlo committed:
reorg panel bodys

Addison-Stavlo committed:
separated inline styles for editor pieces

Addison-Stavlo committed:
added fontSizes to attributes

Addison-Stavlo committed:
added custom sizes to attributes

Addison-Stavlo committed:
color parity between front end and editor

Addison-Stavlo committed:
applied font size selections to front end

Addison-Stavlo committed:
formatting?

Addison-Stavlo committed:
php format

Addison-Stavlo committed:
only show text size pickers for active fields

Addison-Stavlo committed:
add site editor classes for editor styles

Addison-Stavlo committed:
more editor/FE parity

Addison-Stavlo committed:
fontsize fixes in editor

Addison-Stavlo committed:
refactored editor font sizes to usememo

Addison-Stavlo committed:
change to correct classname for text alignment

Addison-Stavlo committed:
add default size for author font

Addison-Stavlo committed:
define some defaults

Addison-Stavlo committed:
weird parity mismatch fix

Addison-Stavlo committed:
remove default fontSize, ruins customFontSize =(

epiqueras reviewed (commented):
Can we use `FONT_SIZE_SUPPORT_KEY = '__experimentalFontSize'` here?

Addison-Stavlo commented:
> Can we use FONT_SIZE_SUPPORT_KEY = '__experimentalFontSize' here?

Im not sure I follow.  Use this where / for what?  Is this some alternative to `withFontSizes` ?

Addison-Stavlo commented:
Oh would that be to support the global styles set fontSize?

This block is odd that with name, byline, and bio, there are 3 distinct sections that should support being different sizes.  Overwriting them all with the same global styles size would severely limit the styling capabilities of the block.

epiqueras commented:
It's a new way of doing font sizes.

Perhaps the secondary text should be sized relative to the "master" font size? Otherwise, it sounds like they should be their own child blocks.

Addison-Stavlo commented:
> It's a new way of doing font sizes.

Ah gotcha.  Im not familiar with it atm, but can take a look tomorrow.  This would require breaking the 3 sections up into child blocks and using this file?
https://github.com/WordPress/gutenberg/blob/32af45067b046e9449789358ad21482d1d3393b5/packages/block-editor/src/hooks/font-size.js#L99

The `withFontSizes` lets us set up all 3 without the need to create child blocks, would this not make more sense currently?  I just kind of took off with the font size tool this had been initialized with.

epiqueras commented:
The Heading block is an example of how to use it.

We need `__experimentalFontSize` to hook into global styles. For now, I recommend the various font sizes are set relative to the font size set for the entire block. Later, if people require it, we can look at creating child blocks or using `__experimentalSelector` to overwrite specific font sizes.

Addison-Stavlo committed:
replace fontSize implementation with __experimentalFontSize

Addison-Stavlo committed:
relative font sizes for other sections

Addison-Stavlo committed:
derp a comment

Addison-Stavlo committed:
updated fontSize being fed to contrastCheckers

Addison-Stavlo committed:
php format

Addison-Stavlo commented:
Interesting!  Thanks for pointing me in the right direction.

I have updated this to use the `__experimentalFontSize` instead, and converted the byline and bio texts to be sized relative to that standard fontSize (0.5em and 0.7em respectively).  There may be more appropriate values to set for these. 🤔 

Im still a bit uncertain regarding global styles:  would this then inherit what is set for global styles by default, and once we specifically save a font size on the block it will use that instead?

epiqueras commented:
>Im still a bit uncertain regarding global styles: would this then inherit what is set for global styles by default, and once we specifically save a font size on the block it will use that instead?

Yes

epiqueras commented on packages/block-library/src/post-author/edit.js:
```diff
@@ -72,15 +83,15 @@ function PostAuthorEdit( {
 			},
 		},
-		[ fontSize.size ]
+		[ contrastCheckFontSize ]
```
We need to change this to `__experimentalUseColors`.

epiqueras replied:
(In a different PR.)

epiqueras commented on packages/block-library/src/post-author/edit.js:
```diff
@@ -205,12 +202,6 @@ function PostAuthorEdit( {
 									className="wp-block-post-author__byline"
 									multiline={ false }
 									placeholder={ __( 'Write byline …' ) }
-									withoutInteractiveFormatting
-									allowedFormats={ [
-										'core/bold',
-										'core/italic',
-										'core/strikethrough',
-									] }
```
What's the reasoning behind this change?

Addison-Stavlo replied:
I figured that if the goal was to have parity with paragraph block's functionalities, that the editable text should be able to have links and the remaining text formatting options.  Removing these removes the restrictions and enables them.

epiqueras replied:
I think these specific restrictions were put in place because other formatting options don't make sense in the context of a byline.

Addison-Stavlo replied:
Why restrict it thought?  If a user wants a word in their byline to be a link, then all of a sudden a link makes sense in the context of a byline for them.  A similar argument for subscript/superscript, etc.

epiqueras commented on packages/block-library/src/post-author/editor.scss:
```diff
@@ -1,12 +1,12 @@
-.block-editor__container .wp-block-post-author {
+.block-editor__container .wp-block-post-author,
+.edit-site-visual-editor .wp-block-post-author {
```
Why is this using `block-editor__container`. I don't think that selector is used anywhere.

Addison-Stavlo replied:
Yeah, `.block-editor__container .wp-block-post-author` was the only selector that was there previously.  Its part of the post editor, but I don't know what it was the class originally put here.  To be safe I just added the `.edit-site-visual-editor .wp-block-post-author` selector as well so the styles would be in the site editor as well.

epiqueras replied:
I don't even see it anywhere in the DOM. We should just use `.wp-block-post-author`.

Addison-Stavlo replied:
> I don't even see it anywhere in the DOM. 

Odd, for me its a child of the .block-editor:
![Screen Shot 2020-06-05 at 6 47 22 PM](https://user-images.githubusercontent.com/28742426/83928264-1a1b1100-a75d-11ea-923b-3d50d53e9a1a.png)

Either way, I think just using `.wp-block-post-author` makes sense.  Updated.

epiqueras commented on packages/block-library/src/post-author/index.php:
```diff
@@ -40,12 +40,6 @@ function post_author_build_css_colors( $attributes ) {
 	$has_named_background_color  = array_key_exists( 'backgroundColor', $attributes );
 	$has_custom_background_color = array_key_exists( 'customBackgroundColor', $attributes );
 
-	// If has background color.
-	if ( $has_custom_background_color || $has_named_background_color ) {
-		// Add has-background-color class.
-		$colors['css_classes'][] = 'has-background-color';
-	}
```
Why is this removed?

Addison-Stavlo replied:
It causes themes to overwrite the font color that gets set and saved.  Themes apply styles to this to selector to set color on the font to ensure it is still readable.  If we are also controlling the font color, then it is unnecessary as well.  Considering it more, we should re-apply this class name, but only when a font color is not set.

Addison-Stavlo replied:
> Considering it more, we should re-apply this class name, but only when a font color is not set.

Or should we?  In that case if someone adds this block and changes the background to a very light color (without changing any font color), a theme would still then force their font color to be really light as well.

Addison-Stavlo replied:
Maybe the best solution is to add this class back in and move the font color styles that get set to inline styles.

epiqueras replied:
We should mirror what the other core blocks are doing.

Addison-Stavlo replied:
Ok, I re-added that class and moved the font color classes/styles into the content div instead.

epiqueras commented on packages/block-library/src/post-author/index.php:
```diff
@@ -72,16 +66,21 @@ function post_author_build_css_font_sizes( $attributes ) {
+		$font_sizes['inline_styles'] = sprintf( 'font-size: %spx;', $attributes['style']['typography']['fontSize'] );
 	}
 
+	$has_named_bio_font_size  = array_key_exists( 'bioFontSize', $attributes );
+	$has_custom_bio_font_size = array_key_exists( 'customBioFontSize', $attributes );
+
```
I don't think this is used.

epiqueras commented on packages/block-library/src/post-author/index.php:
```diff
@@ -117,7 +116,7 @@ function render_block_core_post_author( $attributes, $content, $block ) {
 		isset( $attributes['className'] ) ? array( $attributes['className'] ) : array(),
 		isset( $attributes['itemsJustification'] ) ? array( 'items-justified-' . $attributes['itemsJustification'] ) : array(),
-		isset( $attributes['align'] ) ? array( 'align' . $attributes['align'] ) : array()
+		isset( $attributes['align'] ) ? array( 'has-text-align-' . $attributes['align'] ) : array()
```
Is this change related to this PR?

Addison-Stavlo replied:
That 'align' attribute adjusts the blocks margin on the page.  This attribute corresponds with the text alignment setting so I changed the class name to appropriately reflect that.  

Previously selecting 'align-right' for text alignment would then the block would move to the right of the page, with its text alignment unchanged.

epiqueras replied:
So this was a bug.

Addison-Stavlo committed:
remove unused declarations

Addison-Stavlo committed:
fix font colors with has-background-color class

Addison-Stavlo committed:
removed unnecessary classes from editor styles

Addison-Stavlo committed:
php format?

Addison-Stavlo committed:
update test fixture

Addison-Stavlo committed:
fix editor style regression
